test/functional/JIT_Test/src/jit/test/stringconcat/StringConcatTest.java

```diff
@@ -80,6 +80,16 @@ public void testNullStr_Middle(){
 	private String buildNullStr_Middle(){
 		return Integer.toString(-5) + " is bigger than " + Integer.toString(1) + null + " to a less than sane person";
 	}
+
+	private String concatNonAsciiChar(String s){
+		return new StringBuilder().append(s).append('\u205e').toString();
+	}
+
+	public void testNonAsciiCharConcat(){
+		String str = concatNonAsciiChar("abcd");
+		String str2 = concatNonAsciiChar("abcd");
+		AssertJUnit.assertEquals(str, str2);
+	}
```

Comment:
This is not very robust IMO. Tests should be quite deterministic. This seems to rely on the test being configured to run twice and this method being compiled on the second invocation. If for whatever reason someone were to unknowingly call `concatNonAsciiChar` in some other place in this test suite (for example someone new adding a test similar to yours and thinking to reuse the already existing API) then this test will always pass.

Do you understand the reasoning?

I think a better way to write this test which is more deterministic is:

```
	private String concatChar(String s, char c) {
		return new StringBuilder().append(s).append(c).toString();
	}

	@Test(invocationCount=2)
	public void testStringPeepholeInitStringCharReduction() {
		String str = concatChar("abcd", '\u205e');
		AssertJUnit.assertEquals("abcd\u205e", str);
	}
```

One thing we should check is if @llxia can confirm that `invocationCount` annotation works on an individual test level. I know for a fact it works on the class level, but best to make sure from the experts.

Reply:
I've also added more descriptive names as to what the tests are testing for.

Reply:
yes, `invocationCount` annotation works on an individual test level.
https://www.softpost.org/testng/invocationcount-in-testng/

Reply:
Fixed here: 203a491

Reply:
> yes, `invocationCount` annotation works on an individual test level.
> [softpost.org/testng/invocationcount-in-testng](https://www.softpost.org/testng/invocationcount-in-testng/)

Thanks for the link Lan! I always forget that TestNG is a third party tool. I'll try my best to remember to look up official docs next time as they are readily available.